### PR TITLE
[layer_v2] Merge layer_v2 with main branch

### PR DESCRIPTION
**Pull request by kparichay**
This PR tries to merge layer_v2 with the main branch.
Updates from layer_v2 in this merge:
- All the nntrainer core layer as well as additional layers from applications/unittests have been updated to layer_v2
- All the unittests (except unittest_layers) have been updated for layer_v2 (although unittest_layers has not been updated, layers correctness has been verified using unittest_models)
- All the applications (except unittest of layers in SimpleShot) has been updated for layer_v2
- LayerNode, NetworkGraph, and NeuralNet class have been updated to work with layer_v2
  - All the graph-related properties such as input/output connections, realization properties have been moved to layer_node
  - documentation for this will be added soon
- RNN, LSTM, and GRU layers have been updated to remove some of the temporary creation of tensors
- Pooling layer has been updated to use Tensor via context compared to using vector earlier
- Added mode of execution to be used to internally share the current mode of execution
- layer factory has been removed, layer creation must be done using app_context
- some minor interfaces has been added to LayerInitContext and LayerRunContext
- Extra properties, namely, num_inputs, num_output and in_length have been removed
- backbone has been updated to be trainable by default
- Manager has been updated to support layer_v2 memory support with only shared gradient optimization
- Loss layers have been updated to become independent layers compared to a single layer with internal conditions
- NetworkGraph has been updated to realize loss and properly handle loss as a layer
- Some more minor bug fixes

### CODE REVIEW

**Comment by kparichay** [comment]
Additionally added svace fix for the layer_v2 before merging to main branch.

**Comment by kparichay** [comment]
Please only review the top 7 commits. Other commits are from the layer_v2 branch directly:
1. 1bf9782: merge commit from layer_v2 to main
2. b8e8244: fix svace issues
3. 706c8f2: fix coverity issues
4. 182de84: update dropout layer for v2
5. 6. 7. #1377 commits which have been cherry-picked for layer_v2

File: nntrainer/layers/conv2d_layer.h [hunk 1]
```diff
@@ -35,7 +35,9 @@ class Conv2DLayer : public LayerImpl {
   Conv2DLayer(unsigned int filter_size_ = 0,
               const std::array<unsigned int, CONV2D_DIM> &kernel_size_ = {0, 0},
               const std::array<unsigned int, CONV2D_DIM> &stride_ = {1, 1},
-              const std::array<unsigned int, CONV2D_DIM * 2> &padding_ = {0, 0, 0, 0}) :
+              const std::array<unsigned int, CONV2D_DIM * 2> &padding_ = {0, 0,
+                                                                          0,
+                                                                          0}) :
```
Comment (jijoongmoon) [inline nntrainer/layers/conv2d_layer.h:40]: is it right indentation?

File: nntrainer/layers/layer_node.cpp [hunk 2]
```diff
@@ -116,7 +116,12 @@ int LayerNode::setProperty(std::vector<std::string> properties) {
     auto &ac = nntrainer::AppContext::Global();
     std::unique_ptr<nntrainer::Layer> dlayer =
       ac.createObject<nntrainer::Layer>(TimeDistLayer::type);
-    dynamic_cast<TimeDistLayer *>(dlayer.get())->setDistLayer(std::move(layer));
+    if (dlayer.get() == nullptr)
```
Comment (jijoongmoon) [inline nntrainer/layers/layer_node.cpp:119]: Is there any change dlayer ( ac.createObject()) is nullptr? if it is, definitely createObject need to throw.

File: nntrainer/layers/dropout.cpp [hunk 3]
```diff
@@ -12,60 +12,72 @@
...
+      Tensor &output_ = context.getOutput(i);
+      Tensor &mask_ = context.getTensor(mask_idx[i]);
+
+      mask_ = input_.dropout_mask(rate_);
+      input_.multiply_i(mask_);
 
-      mask_ = input.dropout_mask(rate_);
-      input.multiply_i(mask_);
+      /** @todo: remove below once in_place support is ready from manager */
+      output_.fill(input_);
```
Comment (jijoongmoon) [inline nntrainer/layers/dropout.cpp:52]: good point.

File: nntrainer/layers/dropout.cpp [hunk 4]
```diff
@@ -12,60 +12,72 @@
...
-      deriv.multiply_i(*mask[i].get());
+    for (unsigned int i = 0; i < context.getNumInputs(); ++i) {
+      Tensor &derivative_ = context.getIncomingDerivative(i);
+      Tensor &ret_ = context.getOutgoingDerivative(SINGLE_INOUT_IDX);
+      Tensor &mask_ = context.getTensor(mask_idx[i]);
+
+      derivative_.multiply_i(mask_);
+
+      /** @todo: remove below once in_place support is ready from manager */
+      ret_.fill(derivative_);
```
Comment (jijoongmoon) [inline nntrainer/layers/dropout.cpp:69]: here too.

File: nntrainer/models/neuralnet.cpp [hunk 5]
```diff
@@ -254,7 +234,32 @@ sharedConstTensors NeuralNetwork::forwarding(sharedConstTensors input,
...
+                                             bool training) {
+
+  NNTR_THROW_IF(input[0]->batch() != batch_size ||
+                  (!label.empty() && label[0]->batch() != batch_size),
+                std::logic_error)
+    << "Error: mismatch in batchsize for data and model."
+    << " input_batch: " << input[0]->batch()
+    << " label_batch: " << label[0]->batch() << " target_batch: " << batch_size;
+
+  // std::cerr << "in forward label\n" << *label[0] << "\n";
```
Comment (zhoonit) [inline nntrainer/models/neuralnet.cpp:260]: unintended comment?

Comment (zhoonit) [reply nntrainer/models/neuralnet.cpp:260]: Maybe I should ask @lhs8928?

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/layers/layer_node.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/layer_node.cpp:119]: this is done to resolve coverity issue.

**Review by zhoonit — APPROVED** [review]
LGTM :) Amazing job you've done 🍰

File: nntrainer/layers/conv2d_layer.h (hunk 1 above)
Comment (kparichay) [reply nntrainer/layers/conv2d_layer.h:40]: I find this strange too but clang-format-6.0 updates to this indentation.

Comment (jijoongmoon) [reply nntrainer/layers/conv2d_layer.h:40]: alright.

**Review by lhs8928 — APPROVED** [review]
LGTM except the unintended comment

File: nntrainer/layers/dropout.cpp [hunk 6]
```diff
@@ -12,60 +12,72 @@
...
-int DropOutLayer::setProperty(std::vector<std::string> values) {
-  try {
-    values = loadProperties(values, dropout_rate);
-  } catch (std::invalid_argument &e) {
-    ml_loge("parsing property failed, reason: %s", e.what());
-    return ML_ERROR_INVALID_PARAMETER;
+void DropOutLayer::setProperty(const std::vector<std::string> &values) {
+  auto remain_props = loadProperties(values, dropout_rate);
+  if (!remain_props.empty()) {
+    std::string msg = "[FlattenLayer] Unknown Layer Properties count " +
```
[the remark this answers is not in the document]
Comment (kparichay) [reply nntrainer/layers/dropout.cpp:77]: done.

File: nntrainer/models/neuralnet.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:260]: done.

**Pull request merged**